Proposal: Introduce Dockerfiles for windowsservercore

StefanScherer opened the pull request:
After `golang` now has its first Windows Docker image on [Docker Hub](https://hub.docker.com/_/golang/) it is time for `node` to provide an official Windows Docker image as well.

This PR is a proposal to build and push Windows Docker images for the `windowsservercore` variant which works on Windows 10 + Docker 4 Windows Beta 31++ or Windows Server 2016.

The `test-build.ps1` can be called to build the images. It downloads the ZIP files for Node.js 4.7.0, 6.9.2 and 7.2.1 and builds these Docker images:

- node:4.7.0-windowsservercore
- node:4.7.0-windowsservercore-onbuild
- node:6.9.2-windowsservercore
- node:6.9.2-windowsservercore-onbuild
- node:7.2.1-windowsservercore
- node:7.2.1-windowsservercore-onbuild

In my previous work building a Windows Docker image I used the MSI packages which can be found here https://github.com/StefanScherer/dockerfiles-windows/tree/master/node, but this results in a slightly bigger Docker image as Windows caches the MSI packages in a second place even after removing it in the Dockerfile. So I switched to the ZIP files.

The labels for the Docker images are based on the work for the `golang` Docker images.

See also:
- https://github.com/docker-library/golang/pull/106
- https://hub.docker.com/_/golang/

For information about how to get Docker running on Windows, please see the relevant [Windows Server Quick Start](https://msdn.microsoft.com/en-us/virtualization/windowscontainers/quick_start/quick_start_windows_server) guide provided by Microsoft.

See also #223 for the nanoserver Dockerfile proposal.

Starefossen commented:
This is great! 😄 

Now we just need a CI capable of running Docker for Windows so we can make sure things don't break!

tianon commented on 4.5/windows/windowsservercore/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+ENV NODE_VERSION 4.5.0
+ENV NODE_SHA256 16aab15b29e79746d1bae708f6a5dbed8ef3c87426a9408f7261163d0cda0f56
+
+RUN powershell -Command \
```
You're going to want to include `$ErrorActionPreference = 'Stop';` as the first command in these `powershell` blocks to ensure that failure stops the build appropriately. :+1:

(It's essentially POSIX's `set -e` for PowerShell.)

StefanScherer replied:
Thanks, added that in another commit. I have seen `SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]` in https://github.com/docker-library/golang/blob/ba5f6a10dcc46edeb66377261f873e3b61af36f5/1.7/windows/windowsservercore/Dockerfile#L3 which looks quite elegant, but I don't want to change the default shell for users of this Docker image.

yosifkit replied:
I think the biggest benefit of setting the `SHELL` to `powershell` is that it avoids the unexpected pre-processing of the `RUN` lines by `cmd`, which is the only reason `%NODE_VERSION%` works.  Having one shell interpret the line before it is sent to another shell is one situation I would rather avoid.

StefanScherer replied:
That's a good point.

Starefossen commented:
@tianon can Docker Hub build Docker Images for Windows?

StefanScherer commented:
@Starefossen not yet, also AppVeyor waits for Windows Server 2016 GA.

StefanScherer commented:
I'm more concerned if we need more dev dependencies to make it usable for the community. 

Maybe we should add Python and probanly https://github.com/felixrieseberg/windows-build-tools to allow compiling native node modules as well. 

The golang image without git was also not really usable.

StefanScherer commented:
I just tested the 6.5.0 Dockerfile with another line without any further optimizations

```dockerfile
RUN npm install --global --production windows-build-tools
```

but then the Docker image increases from 7.9 GB to 11.5 GB:

```
C:\Users\vagrant\dockerfiles-windows\node [master ≡ +0 ~1 -0 !]> docker images
REPOSITORY                    TAG                 IMAGE ID            CREATED             SIZE
stefanscherer/node-windows    6.5.0-full          d564a1d10cab        55 seconds ago      11.47 GB
stefanscherer/node-windows    6.5.0               558a2232ccf5        5 days ago          7.887 GB
```

I personally don't have the use case of compiling native node modules and build a Docker image with that for distribution. My workflow would look like this: Run `npm install --production` in AppVeyor in their full installed build images and then `COPY` deploy own code + node_modules into a Docker image (once they have a Docker build image).

friism commented:
@StefanScherer consider using the `SHELL` instruction: https://github.com/Microsoft/sql-server-samples/pull/133/files#diff-3c06805b19714fbfc16e82aba3e542c0R3

StefanScherer commented:
@friism I've tried several times to adapt the Dockerfile to PowerShell without the CMD variable names, but it seems that there is a strange problem using quotes for strings.
As this Dockerfile has the ErrorActionPreference it should be good enough for now.

```
FROM microsoft/windowsservercore
SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
ENV NODE_VERSION 4.6.0

RUN Write-Host "https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NODE_VERSION)-win-x64.zip"
RUN $url = "https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NODE_VERSION)-win-x64.zip"
RUN (New-Object System.Net.WebClient).DownloadFile("https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NODE_VERSION)-win-x64.zip", 'node.zip')
```

shows the correct string after the `Write-Host`, but saving that string as a variable or use it in the DownloadFile function does not work:

```
Step 6/10 : RUN Write-Host "https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NODE_VERSION)-win-x64.zip"
 ---> Running in 96040c5827a5
https://nodejs.org/dist/v4.6.0/node-v4.6.0-win-x64.zip
 ---> 092e42203a2f
Removing intermediate container 96040c5827a5
Step 7/10 : RUN $url = "https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NODE_VERSION)-win-x64.zip"
 ---> Running in 68bcfb0f2eba
https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NODE_VERSION)-win-x6
4.zip : The term 'https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($env:NO
DE_VERSION)-win-x64.zip' is not recognized as the name of a cmdlet, function,
script file, or operable program. Check the spelling of the name, or if a path
was included, verify that the path is correct and try again.
At line:1 char:41
+ ... op'; $url = https://nodejs.org/dist/v$($env:NODE_VERSION)/node-v$($en ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (https://nodejs....ON)-win-x64.z
   ip:String) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : CommandNotFoundException
```

So I keep it as it is right now.

tianon commented:
@StefanScherer in my own testing with PowerShell's string interpolation, it had odd edge cases that made it more trouble than it was worth; I found either `('string' + $var + 'string')` or `('Format {0} string {1}' -f $var1, $var2)` to be much more reliable

StefanScherer commented:
@tianon Thanks. I've added similar workarounds to introduce the `SHELL` instruction. PR rebased and squashed to use latest Node.js versions 4.6 and 6.7.

friism commented:
@chorrell note that there may also (likely) be a `nanoserver` variant alongside `windowsservercore`.

StefanScherer commented:
@chorrell Thanks for #239. Tried something similar today morning, but you know your test scripts better than I.
Your comment about removing the `windows` directory, is it still valid if #223 gets merged in the future to support nanoserver + windowsservercore?
If yes, I'll rebase both PR's, no problem.

chorrell commented:
At the time I thought it might make things simpler to omit the `windows` directory, but I wasn't considering tags and looking at https://github.com/docker-library/golang/pull/106 it makes more sense. So yeah, leave it as is with` windows/nanoserver` and `windows/windowsservercore `. 

We'll want to probably modify the `update.sh` script and add new  `*.template` files for the nanoserver and windowsservercore images at some point. I might also need to do more updates to `test-build.sh` too, but I'll merge #239 first.

StefanScherer commented:
Anything I can do to push this forward a little bit?

chorrell commented:
Was there still an open question about doing CI? I don't have any experience running Docker in Windows and I don't really have a handle on how to test this or what we'd use for automated CI.

StefanScherer commented:
Something like `.travis.yml` + `test_build.sh` for Windows?
This should be possible on AppVeyor soon: `appveyor.yml` + `test_build.ps1`

But I think there is more CI to push to Docker Hub as well, right? Jenkins? So you would need eg. a Windows 2016 Jenkins slave for further steps.

chorrell commented:
The push to docker hub is all manual via a PR e.g., https://github.com/docker-library/official-images/pull/2281

AppVeyor sounds like it might be an option for validating the images then?

StefanScherer commented:
Yes, they plan to setup Win2016 + nanoserver build agents with preloaded base images microsoft/windowsservercore + microsoft/nanoserver.

chorrell commented:
Also, I think a few remaining tasks are:

- Update the Dockerfiles with the latest Node.js versions (v4.6.1, v7.0.0, etc)
- Create `.template` files for the windowsservercore Dockerfiles
- Update the `update.sh` script for updating the Dockerfiles or create a new script just for updating the Windows images.

chorrell commented:
OK, so we probably want to wait until AppVeyor has Windows server 2016. Afaik, the build group doesn't have any 2016 machines in the jenkins/ci setup: https://ci.nodejs.org/computer/

StefanScherer commented:
For example Azure has prebuilt Win2016+Containers including Docker engine preinstalled, there is also an up-to-date "Deploy to Azure" template [here](https://github.com/Microsoft/Virtualization-Documentation/tree/master/windows-server-container-tools/containers-azure-template).
So this should be easy to extend new Jenkins slaves.

I can have a look at the `update.sh`

chorrell commented:
OK, that would be up to the @nodejs/build group to setup I guess.

StefanScherer commented:
The `update.sh` could be enhanced like the one for [docker-library/mongo/update.sh](https://github.com/docker-library/mongo/blob/master/update.sh#L57-L78) fetching the sha256 sums for the Windows Dockerfiles.

StefanScherer commented:
Updated to Node 4.7.0, 6.9.2 and 7.2.1 :-)

computeronix commented:
FYI streamlined the windows side with the msi (of course wouldn't work with nano though)

https://github.com/nodejs/docker-node/pull/297

computeronix commented:
Python has a splitter between Windows and Linux so could be possible to easily implement with the proper tags and automated building mechanism as discussed above
https://store.docker.com/images/1ae86987-df14-4741-9433-d9602a4da995?tab=description

The ones provided by @StefanScherer and myself both are natively supported within Windows Server 2016 Server (and even Azure Server 2016 /w containers ;) )

StefanScherer commented:
FYI, if setting up a Win2016+Docker build agent is too hard, try AppVeyor's new build agent https://github.com/appveyor/ci/issues/1150 using a simple [appveyor.yml](https://github.com/StefanScherer/dockerfiles-windows/blob/master/appveyor.yml) like mine.

StefanScherer commented:
Any progress having an Win 2016 server to provide official Node.js Windows Docker images?

friism commented:
@StefanScherer what about adding appveyor machinery to the PR?

StefanScherer commented:
@friism Thanks, already working on a new PR that introduces both nano + core + appveyor.yml :-)

StefanScherer commented:
Closing, see #362 for a more up to date PR with appveyor.yml
